{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "935400cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import ete3\n",
    "\n",
    "root = '/home/tobiassonva/data/eukgen/'\n",
    "%cd {root}\n",
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ef9ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reformat header by keeping only accession. Save header in separate tsv. Save old header in tsv file for lookup.\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "fastafile = '/home/tobiassonva/data/polvir/polb_reduced_set.faa'\n",
    "seqs = SeqIO.parse(fastafile, 'fasta')\n",
    "print('Writing')\n",
    "with open(fastafile+'.no_header', 'w') as out:\n",
    "    header_dict = {}\n",
    "    for seq in seqs:\n",
    "        try:\n",
    "            header = seq.description.split('|', 1)\n",
    "            #print(header)\n",
    "            header_dict[header[0]] = header[1]\n",
    "        except IndexError:\n",
    "            header_dict[seq.id] = 'no_header'\n",
    "        \n",
    "        seq.description = '' \n",
    "        seq.id = header[0]\n",
    "        SeqIO.write(seq, out, 'fasta')\n",
    "print('Done')\n",
    "\n",
    "with open(fastafile+'.header_mapping', 'w') as out:\n",
    "    for key, value in header_dict.items():\n",
    "        out.write(f'{key}\\t{value}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a409d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reformat header by keeping only accession. Save header in separate tsv. Save old header in tsv file for lookup.\n",
    "#reformmatting from uniprot xx|accession|description\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "fastafile = '/home/tobiassonva/data/databases/human_protein_atlas_localisation/merged_uniprot_acc.fasta'\n",
    "seqs = SeqIO.parse(fastafile, 'fasta')\n",
    "print('Writing')\n",
    "with open(fastafile+'.no_header', 'w') as out:\n",
    "    header_dict = {}\n",
    "    for seq in seqs:\n",
    "        try:\n",
    "            header = seq.description.split('|', 2)\n",
    "            #print(header)\n",
    "            header_dict[header[1]] = header[2]\n",
    "        except IndexError:\n",
    "            header_dict[seq.id] = 'no_header'\n",
    "        \n",
    "        seq.description = '' \n",
    "        seq.id = header[1]\n",
    "        SeqIO.write(seq, out, 'fasta')\n",
    "print('Done')\n",
    "\n",
    "with open(fastafile+'.header_mapping', 'w') as out:\n",
    "    for key, value in header_dict.items():\n",
    "        out.write(f'{key}\\t{value}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d146ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "header = '>sp|Q05655|KPCD_HUMAN Protein kinase C delta type OS=Homo sapiens OX=9606 GN=PRKCD PE=1 SV=2'\n",
    "\n",
    "header.split('|', 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9321fb99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#brute force filter taxa assignment by trying several possible taxonomic rank labels\n",
    "classes = {}\n",
    "for index, row in eukprot_species_taxa.iterrows():\n",
    "    lineage = ete3.NCBITaxa().get_lineage(index)\n",
    "    ranks = ete3.NCBITaxa().get_rank(lineage)\n",
    "    tax_class = -1\n",
    "    \n",
    "    for key, item in ranks.items():\n",
    "        if item == 'class':\n",
    "            tax_class = key\n",
    "            \n",
    "    if tax_class == -1: \n",
    "        #print(index, 'no class')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'phylum':\n",
    "                tax_class = key\n",
    "\n",
    "    \n",
    "    classes[index] = tax_class\n",
    "    if tax_class == -1 :\n",
    "        #print(index, 'no phylum')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'superphylum':\n",
    "                tax_class = key\n",
    "\n",
    "                \n",
    "    if tax_class == -1: \n",
    "        #print(index, 'no superphylum')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'subkingdom':\n",
    "                tax_class = key\n",
    "\n",
    "                \n",
    "    if tax_class == -1: \n",
    "        #print(index, 'no subkingdom')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'kindom':\n",
    "                tax_class = key\n",
    "                \n",
    "    if tax_class == -1: \n",
    "        #print(index, 'no kingdom')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'family':\n",
    "                tax_class = key\n",
    "\n",
    "    if tax_class == -1: \n",
    "        #print(index, 'no family')\n",
    "        for key, item in ranks.items():\n",
    "            if item == 'clade':\n",
    "                tax_class = key\n",
    "    \n",
    "    if tax_class == -1:\n",
    "        print('No taxonomy for ', index)\n",
    "    \n",
    "    classes[index] = tax_class\n",
    "    #print(index, row.species, tax_class)\n",
    "    \n",
    "#final stragllers for eukprot\n",
    "# classes[1246795] = 2686027\n",
    "# classes[652834] = 759891\n",
    "# classes[232287] = 589438\n",
    "# classes[1764928] = 589438\n",
    "# classes[1155914] = 589438"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faee062e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove bad headers from eukprot, slow but only need doing once\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "\n",
    "cleaned_tsv = '/home/tobiassonva/data/eukgen/processing/search/clean_eukprot/clean_eukprot3.tsv'\n",
    "fastafile = '/home/tobiassonva/data/eukgen/core_data/euk72_ep/euk72_ep.fasta' \n",
    "\n",
    "badDB = pd.read_csv(cleaned_tsv, sep='\\t', header=None)\n",
    "bad_accs = badDB[1].unique()\n",
    "\n",
    "seqs = SeqIO.parse(fastafile, 'fasta')\n",
    "\n",
    "print('Writing')\n",
    "with open(fastafile+'.cleaned', 'w') as out:\n",
    "    bads = 0\n",
    "    for seq in seqs:\n",
    "            if seq.id not in bad_accs:\n",
    "                SeqIO.write(seq, out, 'fasta')\n",
    "            else:\n",
    "                bads += 1\n",
    "                bad_id = np.argwhere(bad_accs == seq.id)\n",
    "                bad_accs = np.delete(bad_accs, bad_id)\n",
    "                print(f'found {bads} bad {seq.id}, {bad_accs.shape}')\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9db78b58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d93bb99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in all names and taxa from XML\n",
    "all_assembly_data = []\n",
    "\n",
    "print('Compiling tax for proteins')\n",
    "with open('/home/tobiassonva/data/eukgen/core_data/asgard2023/wolf2024/proteins/assembly_result.xml', 'r') as infile:\n",
    "    for line in infile.readlines():\n",
    "        line = line.strip()\n",
    "        if 'LastMajorRelease' in line:\n",
    "            new_data = [line.strip('<LastMajorReleaseAccession>')[:-2]]\n",
    "        elif '<Taxid>' in line:\n",
    "            new_data.append(line.strip('<Taxid>')[:-2])\n",
    "            all_assembly_data.append(new_data)\n",
    "\n",
    " # format the tax mapping for genome files\n",
    "all_assembly_data = pd.DataFrame(all_assembly_data, columns = ['Assembly', 'orgid'])\n",
    "orgids = ete3.NCBITaxa().get_taxid_translator(all_assembly_data.orgid.unique())\n",
    "all_assembly_data['species'] = [orgids[int(acc)] for acc in all_assembly_data.orgid]\n",
    "all_assembly_data['superkingdom'] = ['Archaea']*all_assembly_data.shape[0]\n",
    "all_assembly_data['class'] = ['Asgard']*all_assembly_data.shape[0]           \n",
    "\n",
    "all_assembly_data\n",
    "\n",
    "print('Expansing to seqs')\n",
    "\n",
    "# iterate over genome files\n",
    "files = subprocess.run(f\"find core_data/asgard2023/wolf2024/proteins/ -name '*.fasta'\", shell=True, text=True, capture_output=True)\n",
    "files = [file for file in files.stdout.split('\\n')][:-1]\n",
    "names = [name.rsplit('/',1)[-1].strip('.fasta') for name in files]\n",
    "\n",
    "all_seq_data = []\n",
    "for n, file in enumerate(files):\n",
    "    print(names[n])\n",
    "    seqs = SeqIO.parse(file, 'fasta')\n",
    "    for seq in seqs:\n",
    "        acc = seq.description.split(' ', 1)[0]\n",
    "        new_data = [acc]\n",
    "        new_data.extend(all_assembly_data[all_assembly_data.Assembly == names[n]][['orgid', 'species', 'superkingdom', 'class']].values[0])\n",
    "        all_seq_data.append(new_data)\n",
    "\n",
    "all_protein_data = pd.DataFrame(all_seq_data, columns = ['acc', 'orgid', 'species', 'superkingdom', 'class'])\n",
    "\n",
    "\n",
    "# iterate over assembly files\n",
    "files = subprocess.run(f\"find core_data/asgard2023/wolf2024/genomic/ -name '*.fasta'\", shell=True, text=True, capture_output=True)\n",
    "files = [file for file in files.stdout.split('\\n')][:-1]\n",
    "names = [name.rsplit('/',1)[-1].rsplit('_', 2)[0] for name in files]\n",
    "\n",
    "print('Compiling tax')\n",
    "all_seq_data = []\n",
    "for n, file in enumerate(files):\n",
    "    seqs = SeqIO.parse(file, 'fasta')\n",
    "    for seq in seqs:\n",
    "        acc = seq.id\n",
    "        if 'MAG:' in seq.description:\n",
    "            species = seq.description.split('MAG: ', 1)[1].rsplit(',', 1)[0].rsplit(' ', 1)[0].rsplit(' ', 1)[0]\n",
    "        elif 'complete genome' in seq.description:\n",
    "            species = seq.description.split(' ',1)[1].rsplit(',', 1)[0].rsplit(' ', 1)[0].rsplit(' ', 1)[0]\n",
    "        else:\n",
    "            print('ERROR: ', seq.description)\n",
    "        all_seq_data.append([acc, species])\n",
    "\n",
    "all_assembly_data = pd.DataFrame(all_seq_data, columns=['acc', 'species'])\n",
    "orgids = ete3.NCBITaxa().get_name_translator(all_assembly_data.species.unique())\n",
    "\n",
    "#manual addendums\n",
    "orgids['Asgard group archaeon isolate'] = [2591838]\n",
    "orgids['Candidatus Asgardarchaeum californiense isolate'] = [3067292]\n",
    "orgids['Candidatus Baldrarchaeota archaeon isolate'] = [2798909]\n",
    "orgids['Candidatus Baldrarchaeum yapensis isolate'] = [2798909]\n",
    "orgids['Candidatus Gerdarchaeota archaeon isolate'] = [2795490]\n",
    "orgids['Candidatus Heimdallarchaeota archaeon isolate'] = [1936272]\n",
    "orgids['Candidatus Heimdallarchaeum aukensis isolate'] = [1936272]\n",
    "orgids['Candidatus Heimdallarchaeum endolithica isolate'] = [1936272]\n",
    "orgids['Candidatus Helarchaeota ASGARD archaeon'] = [2572044]\n",
    "orgids['Candidatus Helarchaeota archaeon isolate'] = [2572044]\n",
    "orgids['Candidatus Hermodarchaeota archaeon isolate'] = [2798910]\n",
    "orgids['Candidatus Hermodarchaeum yapensis isolate'] = [2798910]\n",
    "orgids['Candidatus Hodarchaeota archaeon isolate'] = [2798906]\n",
    "orgids['Candidatus Hodarchaeum mangrovi isolate'] = [2798906]\n",
    "orgids['Candidatus Jordarchaeia archaeon isolate'] = [2823877]\n",
    "orgids['Candidatus Lokiarchaeia archaeon isolate'] = [3049162]\n",
    "orgids['Candidatus Lokiarchaeota archaeon isolate'] = [3049162]\n",
    "orgids['Candidatus Lokiarchaeum ossiferum isolate'] = [3049162]\n",
    "orgids['Candidatus Lokiarchaeum sp.'] = [3049162]\n",
    "orgids['Candidatus Njordarchaeota archaeon isolate'] = [3036961]\n",
    "orgids['Candidatus Njordarchaeum logiae isolate'] = [3036961]\n",
    "orgids['Candidatus Odinarchaeia archaeon isolate'] = [3067281]\n",
    "orgids['Candidatus Odinarchaeum yellowstonii isolate'] = [3067281]\n",
    "orgids['Candidatus Prometheoarchaeum syntrophicum strain'] = [2732160]\n",
    "orgids['Candidatus Sifarchaeia archaeon isolate'] = [2949321]\n",
    "orgids['Candidatus Thorarchaeota archaeon isolate'] = [1706441]\n",
    "orgids['Candidatus Wukongarchaeota archaeon isolate'] = [2798905]\n",
    "orgids['Candidatus Wukongarchaeum yapensis isolate'] = [2798905]\n",
    "\n",
    "all_assembly_data['orgid'] = [orgids[name][0] for name in all_assembly_data.species]\n",
    "all_assembly_data['superkingdom'] = ['Archaea']*all_assembly_data.shape[0]\n",
    "all_assembly_data['class'] = ['Asgard']*all_assembly_data.shape[0]\n",
    "all_assembly_data.set_index('acc', inplace=True)\n",
    "\n",
    "\n",
    "print('Expanding to seqs')\n",
    "# find all accs in prodrigal data\n",
    "file = '/home/tobiassonva/data/eukgen/core_data/asgard2023/wolf2024/genomic/prodrigal.fasta.no_header'\n",
    "seqs = SeqIO.parse(file, 'fasta')\n",
    "accs = [seq.description for seq in seqs]\n",
    "\n",
    "all_seq_data = []\n",
    "for acc in accs:\n",
    "    data = [acc]\n",
    "    data.extend(all_assembly_data.loc[acc.rsplit('_', 1)[0]][['orgid', 'species' ,'superkingdom', 'class']])\n",
    "    all_seq_data.append(data)\n",
    "    \n",
    "all_assembly_data = pd.DataFrame(all_seq_data, columns=['acc', 'orgid', 'species', 'superkingdom', 'class'])\n",
    "print('Done')\n",
    "\n",
    "\n",
    "all_data = pd.concat([all_protein_data, all_assembly_data])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0fc9bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_mapping = pd.read_csv('/home/tobiassonva/data/eukgen/core_data/asgard2023/wolf2024/accession_mapping.tsv', sep ='\\t')\n",
    "orgids = ete3.NCBITaxa().get_name_translator(acc_mapping.Taxonomy.unique())\n",
    "acc_mapping['superkingdom'] = ['Archaea']*acc_mapping.shape[0]\n",
    "acc_mapping['class'] = ['Asgard']*acc_mapping.shape[0]\n",
    "acc_mapping['orgid'] = [orgids[name][0] for name in acc_mapping.Taxonomy]\n",
    "\n",
    "acc_mapping\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea71031",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/home/tobiassonva/data/eukgen/core_data/asgard2023/wolf2024/prodrigal.fasta.no_header'\n",
    "seqs = SeqIO.parse(file, 'fasta')\n",
    "for seq in seqs:\n",
    "    acc = seq.description\n",
    "    data.append([names[n], acc])\n",
    "    print([names[n], acc])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ecef919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# format prok2111_as from merger of prok211 and ettema2023 datasets\n",
    "# use preexisting prok taxonomy and import manuslly curated taxids for ettema2023\n",
    "\n",
    "#load prok header info\n",
    "with open('core_data/taxonomy/prok2111_protein_taxonomy.pkl', 'rb') as infile:\n",
    "    prot_data = pickle.load(infile)\n",
    "\n",
    "#read manual curated asgard header files\n",
    "asgard_seqs = pd.read_csv('core_data/ettema2023/asgard_proteomes2023_header_mapping.tsv', sep ='\\t', names=['acc','header'])\n",
    "asgard_acc_species_map = pd.read_csv('core_data/ettema2023/asgard_acc_species_mapping.tsv', sep='\\t', names=['acc', 'species', 'orgid', 'class'], index_col=0)\n",
    "\n",
    "#format protein_taxonomy table\n",
    "asgard_seqs[['orgid', 'species', 'superkingdom', 'class']] = [[asgard_acc_species_map.loc[acc.split('..')[0]]['orgid'], asgard_acc_species_map.loc[acc.split('..')[0]].species, 'Archaea', 'Asgard'] for acc in asgard_seqs.acc]\n",
    "asgard_seqs.set_index('acc', inplace=True)\n",
    "asgard_seqs.drop('header', axis=1, inplace=True)\n",
    "\n",
    "#write to tsv and pickle\n",
    "asgard_seqs.to_csv('core_data/ettema2023/asgard2023_protein_taxonomy.tsv', sep='\\t')\n",
    "with open('core_data/ettema2023/asgard2023_protein_taxonomy.pkl', 'wb') as pkl:\n",
    "    pickle.dump(asgard_seqs, pkl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b227d3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#additional files\n",
    "\n",
    "merged = pd.concat([prot_data, asgard_seqs])\n",
    "merged.to_csv('core_data/prok2111_as/prok2111_as_protein_taxonomy.tsv', sep='\\t', header=None)\n",
    "taxdb_cols = merged[['orgid']]\n",
    "taxdb_cols.to_csv('core_data/prok2111_as/prok2111.tax', header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0923271",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44251d52",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f318536f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# format taxonomy for final prok2111 and asgard classes with revision\n",
    "data = pd.read_csv('core_data/prok2111/prok2111.tax', sep='\\t', names=['acc', 'orgid'])\n",
    "\n",
    "# get lineages\n",
    "translator =  ete3.NCBITaxa().get_lineage_translator(data.orgid)\n",
    "\n",
    "# Bacteria is 2, only other option is Archaea\n",
    "translator_superkingdom = {key: 'Bacteria' if 2 in value else 'Archaea' for key, value in translator.items()}\n",
    "\n",
    "# add final missing orgids\n",
    "for orgid, values in orgid_addendum.items():\n",
    "    translator_superkingdom[orgid] = values[0]\n",
    "\n",
    "# add data to dataframe\n",
    "data['superkindom'] = [translator_superkingdom[org] for org in data.orgid]\n",
    "\n",
    "\n",
    "#attemt to cocnstruct the \"class\" dataframe\n",
    "translator_class = {}\n",
    "for orgid, lineage in translator.items():\n",
    "    # attempt to grab class from lineage\n",
    "    try:\n",
    "        tax_class_id = [key for key, value in ete3.NCBITaxa().get_rank(lineage).items() if value == 'class']\n",
    "        translator_class[orgid] = tax_class_id[0]\n",
    "    \n",
    "    # if none is found see if its in the addendum\n",
    "    except IndexError:\n",
    "        try:\n",
    "            translator_class[orgid] = class_addendum_ids[lineage[-1]]\n",
    "            \n",
    "        # if not return the orgid for review\n",
    "        except KeyError:\n",
    "            print(orgid, lineage)\n",
    "\n",
    "# add final missing orgids\n",
    "for orgid, values in orgid_addendum.items():\n",
    "    translator_class[orgid] = values[2]\n",
    "    \n",
    "\n",
    "# covert to class names\n",
    "translator_class_name = ete3.NCBITaxa().get_taxid_translator(translator_class.values())\n",
    "\n",
    "# add final missing orgids\n",
    "for orgid, values in orgid_addendum.items():\n",
    "    translator_class_name[orgid] = values[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e80e129c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add and translate classes to dataframe\n",
    "data['class_id'] = [translator_class[orgid] for orgid in data.orgid]\n",
    "data['class'] = [translator_class_name[orgid] if orgid != -1 else 'incertae sedis' for orgid in data.class_id ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce9c1273",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save\n",
    "data.drop(['class_id', 'class_revision'], axis = 1).to_csv('core_data/tmp_tax/prok2111.tax2', header=None, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7594c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "data['class_revision'] = [class_revision[orgid] for orgid in data['class']]\n",
    "\n",
    "#save\n",
    "data.drop(['class', 'class_id'], axis = 1).to_csv('core_data/tmp_tax/prok2111.tax.edited', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "276e4622",
   "metadata": {},
   "outputs": [],
   "source": [
    "# format euk72_ep with final revised taxonomy\n",
    "# format taxonomy for final prok2111 and asgard classes with revision\n",
    "euk_data = pd.read_csv('core_data/euk72_ep/euk72_ep.tax', sep='\\t', names=['acc', 'orgid', 'superkingdom', 'class'])\n",
    "\n",
    "# reassign unknowns\n",
    "unk_ids = euk_data[euk_data['class'] == 'unknown'].orgid.unique()\n",
    "unklist = euk_data[euk_data.orgid.isin(unk_ids)].orgid\n",
    "names = [class_addendum[unk] for unk in unklist]\n",
    "euk_data.loc[euk_data[euk_data.orgid.isin(unk_ids)].index, 'class'] = names\n",
    "\n",
    "# revise classes\n",
    "euk_data['class_revision'] = [class_revision[name] for name in euk_data['class']]\n",
    "\n",
    "euk_data.drop(['class'], axis = 1).to_csv('core_data/tmp_tax/euk.tax.edited', header=None, index=None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67e4704a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class_addendum_ids = {1238993: 544448,\n",
    "1637999: 203494,\n",
    "1779382: 1853220,\n",
    "1540872: -1,\n",
    "1911684: 544448,\n",
    "1911683: 544448,\n",
    "2488809: 203494,\n",
    "2572087: -1,\n",
    "713059: -1,\n",
    "2572089: -1,\n",
    "1930275: 203683,\n",
    "2527975: 203683,\n",
    "2528011: 203683,\n",
    "2528008: 203683,\n",
    "2528012: 203683,\n",
    "2528018: 203683,\n",
    "2527962: 203683,\n",
    "1848255: 1760,\n",
    "1848756: 1760,\n",
    "2725268: 544448,\n",
    "2026780: 203683,\n",
    "2026724: 32061,\n",
    "1932692: 1313211,\n",
    "2755338: 186801,\n",
    "2268202: 649775,\n",
    "713051: -1,\n",
    "2231116: 544448,\n",
    "2886196: 1313211,\n",
    "1871025: 91061,\n",
    "2098: 544448,\n",
    "45363: 544448,\n",
    "92401: 544448,\n",
    "2116: 544448,\n",
    "1048830: 544448,\n",
    "2259673: 1643678,\n",
    "2795383: 1643678,\n",
    "439481: 183967,\n",
    "1229908: 1643678,\n",
    "1959104: 1643678,\n",
    "2508726: 1643678,\n",
    "1470067: 1643678,\n",
    "1229909: 1643678,\n",
    "1470066: 1643678,\n",
    "1580092: 1643678,\n",
    "662947: 544448,\n",
    "662945: 544448,\n",
    "662946: 544448,\n",
    "243273: 544448,\n",
    "663918: 544448,\n",
    "272634: 544448,\n",
    "2104: 544448,\n",
    "1112856: 544448,\n",
    "272633: 544448,\n",
    "710127: 544448,\n",
    "2096: 544448,\n",
    "1159202: 544448,\n",
    "1159204: 544448,\n",
    "1006581: 544448,\n",
    "146919: 1853220,\n",
    "309807: 1853220,\n",
    "761659: 1853220,\n",
    "1159203: 544448,\n",
    "436308: 1643678,\n",
    "1582439: 1643678,\n",
    "1898749: -1,\n",
    "448385: 3031711,\n",
    "56: 3031711,\n",
    "1254432: 3031711,\n",
    "374847: -1,\n",
    "2111: 544448,\n",
    "243272: 544448,\n",
    "1763363: 1117,\n",
    "502025: 3031711,\n",
    "518766: 1853220,\n",
    "762570: 1853220,\n",
    "29549: 1853220,\n",
    "347256: 544448,\n",
    "1267000: 544448,\n",
    "525904: 32061,\n",
    "1263835: 544448,\n",
    "710128: 544448,\n",
    "708616: 544448,\n",
    "1159197: 544448,\n",
    "1159201: 544448,\n",
    "1159199: 544448,\n",
    "1159198: 544448,\n",
    "1159200: 544448,\n",
    "459349: -1,\n",
    "1233262: 544448,\n",
    "1236904: 544448,\n",
    "1213463: 544448,\n",
    "2121: 544448,\n",
    "2124: 544448,\n",
    "2730359: 203494,\n",
    "2796139: 203494,\n",
    "92400: 544448,\n",
    "29559: 544448,\n",
    "1441379: 544448,\n",
    "722438: 544448,\n",
    "2259672: -1,\n",
    "1228987: 1117,\n",
    "52: 3031711,\n",
    "29554: 544448,\n",
    "142649: 544448,\n",
    "1410606: -1,\n",
    "1603555: -1,\n",
    "1846278: -1,\n",
    "927083: 3031711,\n",
    "1263761: 544448,\n",
    "1263760: 544448,\n",
    "1263757: 544448,\n",
    "1263758: 544448,\n",
    "1263756: 544448,\n",
    "1263763: 544448,\n",
    "1263762: 544448,\n",
    "1280940: 544448,\n",
    "243899: 91061,\n",
    "1920749: -1,\n",
    "142651: 544448,\n",
    "2026885: -1,\n",
    "1930593: -1,\n",
    "888845: 3031711,\n",
    "1391654: 3031711,\n",
    "2594042: -1,\n",
    "2715735: -1,\n",
    "281847: 1031332,\n",
    "5722: 5719,\n",
    "1169539: 2605696,\n",
    "5755: 555280,\n",
    "4792: 4762,\n",
    "691883: 2316402,\n",
    "5762: 5752,\n",
    "222440: 2611341,\n",
    "529818: 172820,\n",
    "2903: 2830,\n",
    "31276: 2497438,\n",
    "10228: 3043701,\n",
    "5759: 2605435,\n",
    "5741: 207245,\n",
    "6035: 6029,\n",
    "33067: -1,\n",
    "315: -1,\n",
    "295358: 544448,\n",
    "262719: 544448,\n",
    "262722: 544448,\n",
    "262723: 544448,\n",
    "754503: 544448,\n",
    "33923: 544448,\n",
    "66693: -1,\n",
    "1410606: 1643678,\n",
    "1246955: 544448,\n",
    "67593: 4762,\n",
    "2110: 544448,\n",
    "2113: 544448,\n",
    "2112: 544448,\n",
    "2120: 544448,\n",
    "2951803: 1935183,\n",
    "2725: -1,\n",
    "69011: -1,\n",
    "331677: -1,\n",
    "561169: 2830,\n",
    "1643353: -1,\n",
    "496833: 544448,\n",
    "267748: 544448,\n",
    "38654: 7711,\n",
    "1316930: 544448,\n",
    "2529394: 203494,\n",
    "1579370: -1,\n",
    "1579378: -1,\n",
    "72359: 6029,\n",
    "72586: -1,\n",
    "498716: 544448,\n",
    "171279: 544448,\n",
    "171281: 544448,\n",
    "171282: 544448,\n",
    "171285: 544448,\n",
    "695850: 4762,\n",
    "1318617: 544448,\n",
    "7897: 7711,\n",
    "663362: -1,\n",
    "8467: 7711,\n",
    "8469: 7711,\n",
    "8478: 7711,\n",
    "8479: 7711,\n",
    "8496: 7711,\n",
    "8502: 7711,\n",
    "565575: 544448,\n",
    "42094: 544448,\n",
    "1222016: 91061,\n",
    "632292: 91061,\n",
    "2795018: 2885733,\n",
    "632335: 91061,\n",
    "632348: 91061,\n",
    "370354: 2605435,\n",
    "370355: 2605435,\n",
    "632516: 91061,\n",
    "632518: 91061,\n",
    "403677: 4762,\n",
    "272635: 544448,\n",
    "2665156: -1,\n",
    "273119: 544448,\n",
    "76629: 544448,\n",
    "142650: 544448,\n",
    "77133: -1,\n",
    "1257118: 555280,\n",
    "798570: -1,\n",
    "700510: -1,\n",
    "1618337: -1,\n",
    "1094619: 4762,\n",
    "1619007: -1,\n",
    "1619070: -1,\n",
    "1619077: -1,\n",
    "13735: 7711,\n",
    "79676: -1,\n",
    "505682: 544448,\n",
    "374847: -1,\n",
    "1620412: -1,\n",
    "1882918: 3031711,\n",
    "572263: 544448,\n",
    "48003: 544448,\n",
    "671143: -1,\n",
    "311458: 1643678,\n",
    "114880: 544448,\n",
    "114881: 544448,\n",
    "114885: 544448,\n",
    "1394709: -1,\n",
    "1394711: -1,\n",
    "673860: 183967,\n",
    "608506: 91061,\n",
    "412133: 2611341,\n",
    "1690483: 200643,\n",
    "347257: 544448,\n",
    "1920749: 2885733,\n",
    "2608984: 1042312,\n",
    "414004: 1643678,\n",
    "53558: 544448,\n",
    "184922: 2611341,\n",
    "86660: 544448,\n",
    "578460: 6029,\n",
    "1267001: 544448,\n",
    "907287: 544448,\n",
    "284813: 6029,\n",
    "743966: 544448,\n",
    "743971: 544448,\n",
    "2841263: -1,\n",
    "2841264: -1,\n",
    "907965: 6029,\n",
    "351627: 91061,\n",
    "876142: 6029,\n",
    "289397: 544448,\n",
    "60694: -1,\n",
    "2715735: -1,\n",
    "94835: 7711,\n",
    "29553: 544448,\n",
    "29555: 544448,\n",
    "29556: 544448,\n",
    "1700835: -1,\n",
    "1700836: -1,\n",
    "521460: 91061,\n",
    "1930593: -1,\n",
    "1078905: 1643678,\n",
    "1898108: -1,\n",
    "1603555: 1643678,\n",
    "1178016: 6029,\n",
    "1538547: 1935183,\n",
    "1899017: -1,\n",
    "2259672: 1643678,\n",
    "294381: 2605435,\n",
    "2094: 544448,\n",
    "2099: 544448,\n",
    "2100: 544448,\n",
    "2107: 544448,\n",
    "2109: 544448,\n",
    "2130: 544448,\n",
    "28903: 544448,\n",
    "29561: 544448,\n",
    "38504: 544448,\n",
    "45361: 544448,\n",
    "134821: 544448,\n",
    "171284: 544448,\n",
    "634997: 544448,\n",
    "637387: 544448,\n",
    "872331: 544448,\n",
    "936139: 544448,\n",
    "943945: 544448,\n",
    "956483: 544448,\n",
    "1116211: 544448,\n",
    "1117644: 544448,\n",
    "1118964: 544448,\n",
    "1129369: 544448,\n",
    "1397850: 544448}\n",
    "\n",
    "orgid_addendum = {85643: ['Bacteria', 'Betaproteobacteria', 28216],\n",
    "263820: ['Archaea', 'Thermoplasmata', 183967],\n",
    "228908: ['Archaea', 'Candidatus Nanoarchaeia', 2885752],\n",
    "342610: ['Bacteria', 'Betaproteobacteria', 28216],\n",
    "866775: ['Bacteria', 'Gammaproteobacteria', 1236],\n",
    "1740162: ['Bacteria', 'Betaproteobacteria', 28216],\n",
    "1827146: ['Bacteria', 'Bacilli', 91061],\n",
    "1963032: ['Bacteria', 'Bacilli', 91061],\n",
    "2478912: ['Bacteria', 'Alphaproteobacteria', 28211],\n",
    "2679994: ['Bacteria', 'Gammaproteobacteria', 1236],\n",
    "2795384: ['Bacteria', 'Verrucomicrobiae', 203494]}\n",
    "\n",
    "class_addendum = {1238993: 'Mycoplasmatota',\n",
    "1637999: 'Verrucomicrobiae',\n",
    "1779382: 'Rhodothermota',\n",
    "1540872: 'incertae sedis',\n",
    "1911684: 'Mycoplasmatota',\n",
    "1911683: 'Mycoplasmatota',\n",
    "2488809: 'Verrucomicrobiae',\n",
    "2572087: 'incertae sedis',\n",
    "713059: 'incertae sedis',\n",
    "2572089: 'incertae sedis',\n",
    "1930275: 'Planctomycetia',\n",
    "2527975: 'Planctomycetia',\n",
    "2528011: 'Planctomycetia',\n",
    "2528008: 'Planctomycetia',\n",
    "2528012: 'Planctomycetia',\n",
    "2528018: 'Planctomycetia',\n",
    "2527962: 'Planctomycetia',\n",
    "1848255: 'Actinomycetes',\n",
    "1848756: 'Actinomycetes',\n",
    "2725268: 'Mycoplasmatota',\n",
    "2026780: 'Planctomycetia',\n",
    "2026724: 'Chloroflexia',\n",
    "1932692: 'Lentisphaeria',\n",
    "2755338: 'Clostridia',\n",
    "2268202: 'Synergistia',\n",
    "713051: 'incertae sedis',\n",
    "2231116: 'Mycoplasmatota',\n",
    "2886196: 'Lentisphaeria',\n",
    "1871025: 'Bacilli',\n",
    "2098: 'Mycoplasmatota',\n",
    "45363: 'Mycoplasmatota',\n",
    "92401: 'Mycoplasmatota',\n",
    "2116: 'Mycoplasmatota',\n",
    "1048830: 'Mycoplasmatota',\n",
    "2259673: 'Nitrososphaeria',\n",
    "2795383: 'Nitrososphaeria',\n",
    "439481: 'Thermoplasmata',\n",
    "1229908: 'Nitrososphaeria',\n",
    "1959104: 'Nitrososphaeria',\n",
    "2508726: 'Nitrososphaeria',\n",
    "1470067: 'Nitrososphaeria',\n",
    "1229909: 'Nitrososphaeria',\n",
    "1470066: 'Nitrososphaeria',\n",
    "1580092: 'Nitrososphaeria',\n",
    "662947: 'Mycoplasmatota',\n",
    "662945: 'Mycoplasmatota',\n",
    "662946: 'Mycoplasmatota',\n",
    "243273: 'Mycoplasmatota',\n",
    "663918: 'Mycoplasmatota',\n",
    "272634: 'Mycoplasmatota',\n",
    "2104: 'Mycoplasmatota',\n",
    "1112856: 'Mycoplasmatota',\n",
    "272633: 'Mycoplasmatota',\n",
    "710127: 'Mycoplasmatota',\n",
    "2096: 'Mycoplasmatota',\n",
    "1159202: 'Mycoplasmatota',\n",
    "1159204: 'Mycoplasmatota',\n",
    "1006581: 'Mycoplasmatota',\n",
    "146919: 'Rhodothermota',\n",
    "309807: 'Rhodothermota',\n",
    "761659: 'Rhodothermota',\n",
    "1159203: 'Mycoplasmatota',\n",
    "436308: 'Nitrososphaeria',\n",
    "1582439: 'Nitrososphaeria',\n",
    "1898749: 'incertae sedis',\n",
    "448385: 'Polyangia',\n",
    "56: 'Polyangia',\n",
    "1254432: 'Polyangia',\n",
    "374847: 'incertae sedis',\n",
    "2111: 'Mycoplasmatota',\n",
    "243272: 'Mycoplasmatota',\n",
    "1763363: 'Cyanobacteriota',\n",
    "502025: 'Polyangia',\n",
    "518766: 'Rhodothermota',\n",
    "762570: 'Rhodothermota',\n",
    "29549: 'Rhodothermota',\n",
    "347256: 'Mycoplasmatota',\n",
    "1267000: 'Mycoplasmatota',\n",
    "525904: 'Chloroflexia',\n",
    "1263835: 'Mycoplasmatota',\n",
    "710128: 'Mycoplasmatota',\n",
    "708616: 'Mycoplasmatota',\n",
    "1159197: 'Mycoplasmatota',\n",
    "1159201: 'Mycoplasmatota',\n",
    "1159199: 'Mycoplasmatota',\n",
    "1159198: 'Mycoplasmatota',\n",
    "1159200: 'Mycoplasmatota',\n",
    "459349: 'incertae sedis',\n",
    "1233262: 'Mycoplasmatota',\n",
    "1236904: 'Mycoplasmatota',\n",
    "1213463: 'Mycoplasmatota',\n",
    "2121: 'Mycoplasmatota',\n",
    "2124: 'Mycoplasmatota',\n",
    "2730359: 'Verrucomicrobiae',\n",
    "2796139: 'Verrucomicrobiae',\n",
    "92400: 'Mycoplasmatota',\n",
    "29559: 'Mycoplasmatota',\n",
    "1441379: 'Mycoplasmatota',\n",
    "722438: 'Mycoplasmatota',\n",
    "2259672: 'incertae sedis',\n",
    "1228987: 'Cyanobacteriota',\n",
    "52: 'Polyangia',\n",
    "29554: 'Mycoplasmatota',\n",
    "142649: 'Mycoplasmatota',\n",
    "1410606: 'incertae sedis',\n",
    "1603555: 'incertae sedis',\n",
    "1846278: 'incertae sedis',\n",
    "927083: 'Polyangia',\n",
    "1263761: 'Mycoplasmatota',\n",
    "1263760: 'Mycoplasmatota',\n",
    "1263757: 'Mycoplasmatota',\n",
    "1263758: 'Mycoplasmatota',\n",
    "1263756: 'Mycoplasmatota',\n",
    "1263763: 'Mycoplasmatota',\n",
    "1263762: 'Mycoplasmatota',\n",
    "1280940: 'Mycoplasmatota',\n",
    "243899: 'Bacilli',\n",
    "1920749: 'incertae sedis',\n",
    "142651: 'Mycoplasmatota',\n",
    "2026885: 'incertae sedis',\n",
    "1930593: 'incertae sedis',\n",
    "888845: 'Polyangia',\n",
    "1391654: 'Polyangia',\n",
    "2594042: 'incertae sedis',\n",
    "2715735: 'incertae sedis',\n",
    "281847: 'Cryptomycota',\n",
    "5722: 'Parabasalia',\n",
    "1169539: 'Vitrellaceae',\n",
    "5755: 'Discosea',\n",
    "4792: 'Oomycota',\n",
    "691883: 'Fonticulaceae',\n",
    "5762: 'Heterolobosea',\n",
    "222440: 'Metamonada',\n",
    "529818: 'Apusomonadidae',\n",
    "2903: 'Haptophyta',\n",
    "31276: 'Perkinsozoa',\n",
    "10228: 'Uniplacotomia',\n",
    "5759: 'Evosea',\n",
    "5741: 'Fornicata',\n",
    "6035: 'Microsporidia',\n",
    "33067: 'incertae sedis',\n",
    "315: 'incertae sedis',\n",
    "295358: 'Mycoplasmatota',\n",
    "262719: 'Mycoplasmatota',\n",
    "262722: 'Mycoplasmatota',\n",
    "262723: 'Mycoplasmatota',\n",
    "754503: 'Mycoplasmatota',\n",
    "33923: 'Mycoplasmatota',\n",
    "66693: 'incertae sedis',\n",
    "1410606: 'Nitrososphaeria',\n",
    "1246955: 'Mycoplasmatota',\n",
    "67593: 'Oomycota',\n",
    "2110: 'Mycoplasmatota',\n",
    "2113: 'Mycoplasmatota',\n",
    "2112: 'Mycoplasmatota',\n",
    "2120: 'Mycoplasmatota',\n",
    "2951803: 'Asgard',\n",
    "2725: 'incertae sedis',\n",
    "69011: 'incertae sedis',\n",
    "331677: 'incertae sedis',\n",
    "561169: 'Haptophyta',\n",
    "1643353: 'incertae sedis',\n",
    "496833: 'Mycoplasmatota',\n",
    "267748: 'Mycoplasmatota',\n",
    "38654: 'Chordata',\n",
    "1316930: 'Mycoplasmatota',\n",
    "2529394: 'Verrucomicrobiae',\n",
    "1579370: 'incertae sedis',\n",
    "1579378: 'incertae sedis',\n",
    "72359: 'Microsporidia',\n",
    "72586: 'incertae sedis',\n",
    "498716: 'Mycoplasmatota',\n",
    "171279: 'Mycoplasmatota',\n",
    "171281: 'Mycoplasmatota',\n",
    "171282: 'Mycoplasmatota',\n",
    "171285: 'Mycoplasmatota',\n",
    "695850: 'Oomycota',\n",
    "1318617: 'Mycoplasmatota',\n",
    "7897: 'Chordata',\n",
    "663362: 'incertae sedis',\n",
    "8467: 'Chordata',\n",
    "8469: 'Chordata',\n",
    "8478: 'Chordata',\n",
    "8479: 'Chordata',\n",
    "8496: 'Chordata',\n",
    "8502: 'Chordata',\n",
    "565575: 'Mycoplasmatota',\n",
    "42094: 'Mycoplasmatota',\n",
    "1222016: 'Bacilli',\n",
    "632292: 'Bacilli',\n",
    "2795018: 'Candidatus Micrarchaeia',\n",
    "632335: 'Bacilli',\n",
    "632348: 'Bacilli',\n",
    "370354: 'Evosea',\n",
    "370355: 'Evosea',\n",
    "632516: 'Bacilli',\n",
    "632518: 'Bacilli',\n",
    "403677: 'Oomycota',\n",
    "272635: 'Mycoplasmatota',\n",
    "2665156: 'incertae sedis',\n",
    "273119: 'Mycoplasmatota',\n",
    "76629: 'Mycoplasmatota',\n",
    "142650: 'Mycoplasmatota',\n",
    "77133: 'incertae sedis',\n",
    "1257118: 'Discosea',\n",
    "798570: 'incertae sedis',\n",
    "700510: 'incertae sedis',\n",
    "1618337: 'incertae sedis',\n",
    "1094619: 'Oomycota',\n",
    "1619007: 'incertae sedis',\n",
    "1619070: 'incertae sedis',\n",
    "1619077: 'incertae sedis',\n",
    "13735: 'Chordata',\n",
    "79676: 'incertae sedis',\n",
    "505682: 'Mycoplasmatota',\n",
    "374847: 'incertae sedis',\n",
    "1620412: 'incertae sedis',\n",
    "1882918: 'Polyangia',\n",
    "572263: 'Mycoplasmatota',\n",
    "48003: 'Mycoplasmatota',\n",
    "671143: 'incertae sedis',\n",
    "311458: 'Nitrososphaeria',\n",
    "114880: 'Mycoplasmatota',\n",
    "114881: 'Mycoplasmatota',\n",
    "114885: 'Mycoplasmatota',\n",
    "1394709: 'incertae sedis',\n",
    "1394711: 'incertae sedis',\n",
    "673860: 'Thermoplasmata',\n",
    "608506: 'Bacilli',\n",
    "412133: 'Metamonada',\n",
    "1690483: 'Bacteroidia',\n",
    "347257: 'Mycoplasmatota',\n",
    "1920749: 'Candidatus Micrarchaeia',\n",
    "2608984: 'Armatimonadia',\n",
    "414004: 'Nitrososphaeria',\n",
    "53558: 'Mycoplasmatota',\n",
    "184922: 'Metamonada',\n",
    "86660: 'Mycoplasmatota',\n",
    "578460: 'Microsporidia',\n",
    "1267001: 'Mycoplasmatota',\n",
    "907287: 'Mycoplasmatota',\n",
    "284813: 'Microsporidia',\n",
    "743966: 'Mycoplasmatota',\n",
    "743971: 'Mycoplasmatota',\n",
    "2841263: 'incertae sedis',\n",
    "2841264: 'incertae sedis',\n",
    "907965: 'Microsporidia',\n",
    "351627: 'Bacilli',\n",
    "876142: 'Microsporidia',\n",
    "289397: 'Mycoplasmatota',\n",
    "60694: 'incertae sedis',\n",
    "2715735: 'incertae sedis',\n",
    "94835: 'Chordata',\n",
    "29553: 'Mycoplasmatota',\n",
    "29555: 'Mycoplasmatota',\n",
    "29556: 'Mycoplasmatota',\n",
    "1700835: 'incertae sedis',\n",
    "1700836: 'incertae sedis',\n",
    "521460: 'Bacilli',\n",
    "1930593: 'incertae sedis',\n",
    "1078905: 'Nitrososphaeria',\n",
    "1898108: 'incertae sedis',\n",
    "1603555: 'Nitrososphaeria',\n",
    "1178016: 'Microsporidia',\n",
    "1538547: 'Asgard',\n",
    "1899017: 'incertae sedis',\n",
    "2259672: 'Nitrososphaeria',\n",
    "294381: 'Evosea',\n",
    "2094: 'Mycoplasmatota',\n",
    "2099: 'Mycoplasmatota',\n",
    "2100: 'Mycoplasmatota',\n",
    "2107: 'Mycoplasmatota',\n",
    "2109: 'Mycoplasmatota',\n",
    "2130: 'Mycoplasmatota',\n",
    "28903: 'Mycoplasmatota',\n",
    "29561: 'Mycoplasmatota',\n",
    "38504: 'Mycoplasmatota',\n",
    "45361: 'Mycoplasmatota',\n",
    "134821: 'Mycoplasmatota',\n",
    "171284: 'Mycoplasmatota',\n",
    "634997: 'Mycoplasmatota',\n",
    "637387: 'Mycoplasmatota',\n",
    "872331: 'Mycoplasmatota',\n",
    "936139: 'Mycoplasmatota',\n",
    "943945: 'Mycoplasmatota',\n",
    "956483: 'Mycoplasmatota',\n",
    "1116211: 'Mycoplasmatota',\n",
    "1117644: 'Mycoplasmatota',\n",
    "1118964: 'Mycoplasmatota',\n",
    "1129369: 'Mycoplasmatota',\n",
    "1397850: 'Mycoplasmatota'}\n",
    "\n",
    "class_revision = {'Gammaproteobacteria': 'Gammaproteobacteria',\n",
    "'Actinomycetes': 'Actinomycetota',\n",
    "'Alphaproteobacteria': 'Alphaproteobacteria',\n",
    "'Betaproteobacteria': 'Betaproteobacteria',\n",
    "'Deltaproteobacteria': 'Deltaproteobacteria',\n",
    "'Mollicutes': 'Mycoplasmatota',\n",
    "'Mycoplasmatota':'Mycoplasmatota',\n",
    "'incertae sedis': 'DELETE',\n",
    "'Korarchaeia': 'TACK group',\n",
    "'Myxococcia': 'Myxococcota',\n",
    "'Bipolaricaulia': 'DELETE',\n",
    "'Chloroflexia': 'Chloroflexota',\n",
    "'Thermodesulfobacteria': 'Thermodesulfobacteriota',\n",
    "'Deferribacteres': 'DELETE',\n",
    "'Acidimicrobiia': 'Actinomycetota',\n",
    "'Rubrobacteria': 'Actinomycetota',\n",
    "'Coriobacteriia': 'Actinomycetota',\n",
    "'Bacilli': 'Bacillota',\n",
    "'Flavobacteriia': 'FCB group',\n",
    "'Sphingobacteriia': 'FCB group',\n",
    "'Chrysiogenetes': 'DELETE',\n",
    "'Spartobacteria': 'PVC group',\n",
    "'Thermoprotei': 'TACK group',\n",
    "'Methanobacteria': 'Euryarchaeota',\n",
    "'Methanococci': 'Euryarchaeota',\n",
    "'Halobacteria': 'Euryarchaeota',\n",
    "'Thermoplasmata': 'Thermoplasmata',\n",
    "'Thermococci': 'Euryarchaeota',\n",
    "'Archaeoglobi': 'Euryarchaeota',\n",
    "'Methanopyri': 'Euryarchaeota',\n",
    "'Clostridia': 'Bacillota',\n",
    "'Aquificae': 'DELETE',\n",
    "'Thermotogae': 'Thermotogae',\n",
    "'Rhodothermia': 'FCB group',\n",
    "'Deinococci': 'DELETE',\n",
    "'Thermomicrobia': 'DELETE',\n",
    "'Chlorobiia': 'FCB group',\n",
    "'Bacteroidia': 'FCB group',\n",
    "'Dictyoglomia': 'DELETE',\n",
    "'Fusobacteriia': 'DELETE',\n",
    "'Verrucomicrobiae': 'PVC group',\n",
    "'Planctomycetia': 'PVC group',\n",
    "'Spirochaetia': 'DELETE',\n",
    "'Nitrospiria': 'DELETE',\n",
    "'Chlamydiia': 'Chlamydiia',\n",
    "'Fibrobacteria': 'FCB group',\n",
    "'Terriglobia': 'Acidobacteriota',\n",
    "'Gemmatimonadetes': 'FCB group',\n",
    "'Methanomicrobia': 'Euryarchaeota',\n",
    "'Anaerolineae': 'Chloroflexota',\n",
    "'Dehalococcoidia': 'Chloroflexota',\n",
    "'Gracilibacteria': 'DELETE',\n",
    "'Ktedonobacteria': 'Chloroflexota',\n",
    "'Opitutae': 'PVC group',\n",
    "'Endomicrobiia': 'DELETE',\n",
    "'Caldilineae': 'Chloroflexota',\n",
    "'Erysipelotrichia': 'Bacillota',\n",
    "'Holophagae': 'Acidobacteriota',\n",
    "'Zetaproteobacteria': 'DELETE',\n",
    "'Elusimicrobia': 'DELETE',\n",
    "'Synergistia': 'DELETE',\n",
    "'Phycisphaerae': 'PVC group',\n",
    "'Caldisericia': 'DELETE',\n",
    "'Cytophagia': 'FCB group',\n",
    "'Ignavibacteria': 'FCB group',\n",
    "'Nitriliruptoria': 'Actinomycetota',\n",
    "'Negativicutes': 'Bacillota',\n",
    "'Nanohaloarchaea': 'Euryarchaeota',\n",
    "'Lentisphaeria': 'PVC group',\n",
    "'Ardenticatenia': 'Chloroflexota',\n",
    "'Thermoleophilia': 'Actinomycetota',\n",
    "'Oligoflexia': 'DELETE',\n",
    "'Blastocatellia': 'Acidobacteriota',\n",
    "'Nitrososphaeria': 'TACK group',\n",
    "'Fimbriimonadia': 'DELETE',\n",
    "'Limnochordia': 'Bacillota',\n",
    "'Tissierellia': 'Bacillota',\n",
    "'Acidithiobacillia': 'DELETE',\n",
    "'Vicinamibacteria': 'Acidobacteriota',\n",
    "'Rhodothermota': 'FCB group',\n",
    "'Balneolia': 'FCB group',\n",
    "'Chitinophagia': 'FCB group',\n",
    "'Candidatus Izimaplasma': 'Mycoplasmatota',\n",
    "'Kiritimatiellia': 'PVC group',\n",
    "'Asgard': 'Asgard',\n",
    "'Saprospiria': 'FCB group',\n",
    "'Methylacidiphilae': 'PVC group',\n",
    "'Calditrichia': 'DELETE',\n",
    "'Hydrogenophilia': 'DELETE',\n",
    "'Candidatus Saccharimonadia': 'DELETE',\n",
    "'Saccharimonadia': 'DELETE',\n",
    "'Coprothermobacteria': 'DELETE',\n",
    "'Candidatus Thermofonsia': 'Chloroflexota',\n",
    "'Candidatus Babeliae': 'DELETE',\n",
    "'Candidatus Brocadiia': 'PVC group',\n",
    "'Tepidiformia': 'Chloroflexota',\n",
    "'Conexivisphaeria': 'TACK group',\n",
    "'Thermodesulfovibrionia': 'DELETE',\n",
    "'Atribacteria': 'DELETE',\n",
    "'Candidatus Nanohalobia': 'DPANN group',\n",
    "'Candidatus Micrarchaeia': 'DPANN group',\n",
    "'Micrarchaeia': 'DPANN group',\n",
    "'Candidatus Nanoarchaeia': 'DPANN group',\n",
    "'Tichowtungiia': 'PVC group',\n",
    "'Syntrophobacteria': 'Thermodesulfobacteriota',\n",
    "'Desulfobacteria': 'Thermodesulfobacteriota',\n",
    "'Cyanophyceae': 'Cyanobacteriota',\n",
    "'Bdellovibrionia': 'DELETE',\n",
    "'Bacteriovoracia': 'Bdellovibrionota',\n",
    "'Desulfovibrionia': 'Thermodesulfobacteriota',\n",
    "'Desulfobulbia': 'Thermodesulfobacteriota',\n",
    "'Desulfarculia': 'Thermodesulfobacteriota',\n",
    "'Desulfobaccia': 'Thermodesulfobacteriota',\n",
    "'Syntrophia': 'Thermodesulfobacteriota',\n",
    "'Desulfomonilia': 'Thermodesulfobacteriota',\n",
    "'Desulfuromonadia': 'Thermodesulfobacteriota',\n",
    "'Polyangia': 'Myxococcota',\n",
    "'Epsilonproteobacteria': 'Campylobacterota',\n",
    "'Desulfurellia': 'Campylobacterota',\n",
    "'Bangiophyceae': 'Rhodophyta',\n",
    "'Florideophyceae': 'Rhodophyta',\n",
    "'Chrysophyceae': 'Ochrophyta',\n",
    "'Haptophyta': 'Haptista',\n",
    "'Xanthophyceae': 'Ochrophyta',\n",
    "'Dinophyceae': 'Dinophyceae',\n",
    "'Phaeophyceae': 'Ochrophyta',\n",
    "'Cryptophyceae': 'Cryptophyceae',\n",
    "'Euglenida': 'Euglenida',\n",
    "'Chlorophyta': 'Chlorophyta',\n",
    "'Chlorophyceae': 'Chlorophyta',\n",
    "'Bryopsida': 'Streptophyta',\n",
    "'Gnetopsida': 'Streptophyta',\n",
    "'Magnoliopsida': 'Streptophyta',\n",
    "'Oomycota': 'Oomycota',\n",
    "'Saccharomycetes': 'Ascomycota',\n",
    "'Ustilaginomycetes': 'Basidiomycota',\n",
    "'Kinetoplastea': 'Euglenida',\n",
    "'Parabasalia': 'Metamonada',\n",
    "'Eustigmatophyceae': 'Ochrophyta',\n",
    "'Heterolobosea': 'Heterolobosea',\n",
    "'Colpodea': 'Ciliophora',\n",
    "'Litostomatea': 'Ciliophora',\n",
    "'Prostomatea': 'Ciliophora',\n",
    "'Nassophorea': 'Ciliophora',\n",
    "'Oligohymenophorea': 'Ciliophora',\n",
    "'Microsporidia': 'Microsporidia',\n",
    "'Demospongiae': 'Metazoa',\n",
    "'Hydrozoa': 'Metazoa',\n",
    "'Anthozoa': 'Metazoa',\n",
    "'Scyphozoa': 'Metazoa',\n",
    "'Trematoda': 'Metazoa',\n",
    "'Cestoda': 'Metazoa',\n",
    "'Polychaeta': 'Metazoa',\n",
    "'Gastropoda': 'Metazoa',\n",
    "'Bivalvia': 'Metazoa',\n",
    "'Cephalopoda': 'Metazoa',\n",
    "'Branchiopoda': 'Metazoa',\n",
    "'Arachnida': 'Metazoa',\n",
    "'Chilopoda': 'Metazoa',\n",
    "'Echinoidea': 'Metazoa',\n",
    "'Chordata': 'Metazoa',\n",
    "'Ascidiacea': 'Metazoa',\n",
    "'Chondrichthyes': 'Metazoa',\n",
    "'Amphibia': 'Metazoa',\n",
    "'Lepidosauria': 'Metazoa',\n",
    "'Aves': 'Metazoa',\n",
    "'Calcarea': 'Metazoa',\n",
    "'Choanoflagellata': 'Choanoflagellata',\n",
    "'Foraminifera': 'Rhizaria',\n",
    "'Chlorarachniophyceae': 'Rhizaria',\n",
    "'Ginkgoopsida': 'Streptophyta',\n",
    "'Ulvophyceae': 'Chlorophyta',\n",
    "'Opisthokonta': 'Opisthokonta',\n",
    "'Spirotrichea': 'Ciliophora',\n",
    "'Coscinodiscophyceae': 'Ochrophyta',\n",
    "'Bacillariophyceae': 'Ochrophyta',\n",
    "'Fragilariophyceae': 'Ochrophyta',\n",
    "'Synurophyceae': 'Ochrophyta',\n",
    "'Pelagophyceae': 'Ochrophyta',\n",
    "'Phyllopharyngea': 'Ciliophora',\n",
    "'Glaucocystophyceae': 'Glaucocystophyceae',\n",
    "'Raphidophyceae': 'Ochrophyta',\n",
    "'Dictyochophyceae': 'Ochrophyta',\n",
    "'Mammalia': 'Metazoa',\n",
    "'Clitellata': 'Metazoa',\n",
    "'Insecta': 'Metazoa',\n",
    "'Pinopsida': 'Streptophyta',\n",
    "'Hexactinellida': 'Metazoa',\n",
    "'Acantharea': 'Rhizaria',\n",
    "'Polycystinea': 'Rhizaria',\n",
    "'Hexanauplia': 'Metazoa',\n",
    "'Trebouxiophyceae': 'Chlorophyta',\n",
    "'Homoscleromorpha': 'Metazoa',\n",
    "'Ancyromonadidae': 'Ancyromonadidae',\n",
    "'Bolidophyceae': 'Ochrophyta',\n",
    "'Mesostigmatophyceae': 'Streptophyta',\n",
    "'Lingulata': 'Metazoa',\n",
    "'Hyperoartia': 'Metazoa',\n",
    "'Chromadorea': 'Metazoa',\n",
    "'Ichthyosporea': 'Ichthyosporea',\n",
    "'Zygnemophyceae': 'Streptophyta',\n",
    "'Chlorokybophyceae': 'Streptophyta',\n",
    "'Klebsormidiophyceae': 'Streptophyta',\n",
    "'Malawimonadidae': 'Malawimonadidae',\n",
    "'Cercozoa': 'Rhizaria',\n",
    "'Tentaculata': 'Metazoa',\n",
    "'Eumycetozoa': 'Amoebozoa',\n",
    "'Jakobidae': 'Jakobida',\n",
    "'Dothideomycetes': 'Ascomycota',\n",
    "'Eurotiomycetes': 'Ascomycota',\n",
    "'Lecanoromycetes': 'Ascomycota',\n",
    "'Pezizomycetes': 'Ascomycota',\n",
    "'Sordariomycetes': 'Ascomycota',\n",
    "'Schizosaccharomycetes': 'Ascomycota',\n",
    "'Taphrinomycetes': 'Ascomycota',\n",
    "'Neolectomycetes': 'Ascomycota',\n",
    "'Tremellomycetes': 'Basidiomycota',\n",
    "'Agaricomycetes': 'Basidiomycota',\n",
    "'Pinguiophyceae': 'Ochrophyta',\n",
    "'Microbotryomycetes': 'Basidiomycota',\n",
    "'Apusomonadidae': 'Apusozoa',\n",
    "'Colpodellaceae': 'Colpodellida',\n",
    "'Actinopteri': 'Metazoa',\n",
    "'Marchantiopsida': 'Streptophyta',\n",
    "'Collodictyonidae': 'CRuMs',\n",
    "'Diplonemea': 'Euglenida',\n",
    "'Centroplasthelida': 'Haptista',\n",
    "'Heterotrichea': 'Ciliophora',\n",
    "'Fornicata': 'Metamonada',\n",
    "'Glomeromycetes': 'Fungi incertae sedis',\n",
    "'Polypodiopsida': 'Streptophyta',\n",
    "'Coleochaetophyceae': 'Streptophyta',\n",
    "'Charophyceae': 'Streptophyta',\n",
    "'Anthocerotopsida': 'Streptophyta',\n",
    "'Picozoa': 'Eukaryota incertae sedis',\n",
    "'Aconoidasida': 'Apicomplexa',\n",
    "'Wallemiomycetes': 'Basidiomycota',\n",
    "'Cystobasidiomycetes': 'Basidiomycota',\n",
    "'Stylonematophyceae': 'Rhodophyta',\n",
    "'Chytridiomycetes': 'Fungi incertae sedis',\n",
    "'Monoblepharidomycetes': 'Fungi incertae sedis',\n",
    "'Neocallimastigomycetes': 'Fungi incertae sedis',\n",
    "'Blastocladiomycetes': 'Fungi incertae sedis',\n",
    "'Amoebozoa': 'Amoebozoa',\n",
    "'Discosea': 'Amoebozoa',\n",
    "'Histionidae': 'Jakobida',\n",
    "'Synchromophyceae': 'Ochrophyta',\n",
    "'Telonemida': 'Telonemida',\n",
    "'Mediophyceae': 'Ochrophyta',\n",
    "'Armophorea': 'Ciliophora',\n",
    "'Rhodellophyceae': 'Rhodophyta',\n",
    "'Palpitomonas': 'Eukaryota incertae sedis',\n",
    "'Compsopogonophyceae': 'Rhodophyta',\n",
    "'Thecofilosea': 'Rhizaria',\n",
    "'Cryptomycota': 'Fungi incertae sedis',\n",
    "'Mamiellophyceae': 'Chlorophyta',\n",
    "'Tsukubamonadidae': 'Tsukubamonadidae',\n",
    "'Xylonomycetes': 'Ascomycota',\n",
    "'Mantamonadidae': 'CRuMs',\n",
    "'Nephroselmidophyceae': 'Chlorophyta',\n",
    "'Plagiopylea': 'Ciliophora',\n",
    "'Conoidasida': 'Apicomplexa',\n",
    "'Planomonadidae': 'Ancyromonadidae',\n",
    "'Picomonadea': 'Eukaryota incertae sedis',\n",
    "'Basidiobolomycetes': 'Fungi incertae sedis',\n",
    "'Entomophthoromycetes': 'Fungi incertae sedis',\n",
    "'Breviatea': 'Breviatea',\n",
    "'Flabellinia': 'Amoebozoa',\n",
    "'Colponemidia': 'Colponemidia',\n",
    "'Aphelidea': 'Aphelidea',\n",
    "'Lycopodiopsida': 'Streptophyta',\n",
    "'Chlorodendrophyceae': 'Chlorophyta',\n",
    "'Malasseziomycetes': 'Basidiomycota',\n",
    "'Imbricatea': 'Rhizaria',\n",
    "'Ascetosporea': 'Rhizaria',\n",
    "'Palmophyllophyceae': 'Chlorophyta',\n",
    "'Mucoromycetes': 'Fungi incertae sedis',\n",
    "'Mortierellomycetes': 'Fungi incertae sedis',\n",
    "'Kickxellomycetes': 'Fungi incertae sedis',\n",
    "'Dimargaritomycetes': 'Fungi incertae sedis',\n",
    "'Harpellomycetes': 'Fungi incertae sedis',\n",
    "'Zoopagomycetes': 'Fungi incertae sedis',\n",
    "'Chloropicophyceae': 'Chlorophyta',\n",
    "'Fonticulaceae': 'Fonticulaceae',\n",
    "'Hemimastigophora': 'Hemimastigophora',\n",
    "'Perkinsozoa': 'Perkinsozoa',\n",
    "'Rhodelphea': 'Rhodelphea',\n",
    "'Endomyxa': 'Rhizaria',\n",
    "'Variosea': 'Amoebozoa',\n",
    "'Evosea': 'Amoebozoa',\n",
    "'Vitrellaceae': 'Colpodellida',\n",
    "'Chromeraceae': 'Colpodellida',\n",
    "'Alphamonaceae': 'Colpodellida',\n",
    "'CRuMs': 'CRuMs',\n",
    "'Metamonada': 'Metamonada',\n",
    "'Preaxostyla': 'Metamonada',\n",
    "'Echinamoebida': 'Amoebozoa',\n",
    "'Elardia': 'Amoebozoa',\n",
    "'Leptocardii': 'Metazoa',\n",
    "'Andaluciidae': 'Jakobida',\n",
    "'Stygiellidae': 'Jakobida',\n",
    "'Bigyra': 'Bigyra',\n",
    "'Ancyromonadida': 'Ancyromonadidae',\n",
    "'Filasterea': 'Filasterea',\n",
    "'Picocystophyceae': 'Chlorophyta',\n",
    "'Sar': 'Sar',\n",
    "'Pyramimonadophyceae': 'Chlorophyta',\n",
    "'Muranotrichea': 'Ciliophora',\n",
    "'Phytomyxea': 'Rhizaria',\n",
    "'Prasinodermophyceae': 'Prasinodermophyceae',\n",
    "'Sanchytriomycetes': 'Fungi incertae sedis',\n",
    "'Marosporida': 'Apicomplexa',\n",
    "'Nebulidea': 'Nebulidea',\n",
    "'Uniplacotomia': 'Metazoa',\n",
    "'Candidatus Cloacimonadia': 'FCB group',\n",
    "'Candidatus Lokiarchaeia': 'Asgard',\n",
    "'Candidatus Nanohaloarchaea': 'Euryarchaeota',\n",
    "'Cyanobacteriota': 'Cyanobacteriota'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2217002",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_annot(file):\n",
    "    data = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "    \n",
    "    data[['label_type', 'predicted_label']] = [label.split(':') for label in data.predicted_label]\n",
    "\n",
    "    data_pfam = data[data.label_type == 'Pfam']\n",
    "    data_EC = data[data.label_type == 'EC']\n",
    "    data_GO = data[data.label_type == 'GO']\n",
    "    \n",
    "    data_EC = data_EC.reset_index().drop_duplicates('sequence_name', keep='last').set_index('sequence_name')\n",
    "    data_EC['defined_EC'] = [True if label[-1] != '-' else False for label in data_EC.predicted_label]\n",
    "    \n",
    "    return data_pfam, data_EC, data_GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6445ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prok_labels = pd.read_csv('core_data/prok2111_as/prok2111_as.fasta.header_mapping', sep='\\t', index_col=0, header=None)\n",
    "euk_labels = labels = pd.read_csv('core_data/euk72_ep/euk72_ep.fasta.header_mapping', sep='\\t', index_col=0, header=None)\n",
    "\n",
    "labels= pd.concat([prok_labels, euk_labels])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1f927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "prok_labels = prok_labels.columns = ['acc', 'desc']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73941d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels.columns = ['desc']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "792fbbff",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dabce1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "euk_labels. = ['acc', 'desc']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebd20437",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_pfam, data_EC, data_GO = parse_annot('/data/luojaa/proteinfer/annot_microcosms/glycolysis.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b146fa65",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC[data_EC.defined_EC]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd3bfd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('/data/luojaa/proteinfer/annot_microcosms/glycolysis.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6de4138b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.index.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6a49c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_GO.description.value_counts()[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58c8b47e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC[(data_EC.defined_EC) & (data_EC.confidence > 0.9)].description.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2fb56d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC[(data_EC.defined_EC) & (data_EC.confidence > 0.9)].description.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1324c025",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = data_EC[data_EC.description == 'Ferredoxin--NADP(+) reductase.'].index\n",
    "labels.loc[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21cfefb6",
   "metadata": {},
   "outputs": [],
   "source": [
    " data_EC[data_EC.description == 'Ferredoxin--NADP(+) reductase.']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "806d33a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "prok_i = data_EC[(data_EC.defined_EC) & (data_EC.confidence > 0.9)].index[10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aebb23c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC.loc[prok_i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0776f3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(*data_EC.loc[prok_i].index.values, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c1ee065",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels.loc[prok_i].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "305b36d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC.iloc[:31].reset_index()#.drop_duplicates('predicted_label', keep='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71dbe6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_EC.iloc[:31].reset_index().drop_duplicates('sequence_name', keep='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c231096",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[(data.defined_EC) & (data.label_type == 'EC') & (data.confidence > 0.95)]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46fb9c25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21637f82",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b1f5fc9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe23ce5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108661b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62810402",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.9",
   "language": "python",
   "name": "py3.9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
